### ISSUE TITLE
Provide "committed resource" semantics in HAL

### ISSUE THREAD

**Issue by kvark** [gfx-rs/gfx#2511]
Primary motivation is getting advantage of [vk-mem](https://crates.io/crates/vk-mem) by @gwihlidal in our Vulkan backend. The new API entry points for buffers and images could return BoundXxx right away.

Implementing in D3D11/D3D12 and Metal is also going to be straightforward.

**Comment by grovesNL** [comment]
Is there a way we could make these optional so users could opt-in? i.e. thinking about the rendy-memory case where memory allocation is managed at the HAL abstraction level, not per backend.

Maybe we could have an optional feature and expose entry points through a HAL extension?

**Comment by grovesNL** [comment]
Right, it seems a bit unfortunate to always include the additional dependency (mostly since it requires building a C++ project).

**Comment by kvark** [comment]
Now that I think of it, we don't *have to* include this dependency. We'll get the cleanest solution by having `vk-mem` dependency as optional (off by default), and just allocate a memory chunk per object (ourselves) if those new entry points are used without the dependency enabled. Hopefully, nobody would depend on this behavior in critical code ;)
